File: src/web/Controllers/OptionsController.cs
using System.Threading.Tasks;
using core.fs.Options;
using core.Shared;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using web.Utils;

namespace web.Controllers
{
    [ApiController]
    [Authorize]
    [Route("api/[controller]")]
    public class OptionsController(OptionsHandler handler) : ControllerBase
    {
        [HttpGet("pricing")]
        public Task<ActionResult> GetOptionPricing([FromQuery] string symbol) =>
            this.OkOrError(
                handler.Handle(
                    new OptionPricingQuery(
                        User.Identifier(), OptionTicker.NewOptionTicker(symbol)
                    )
                )
            );
        
        [HttpGet("chain/{ticker}")]
        public Task<ActionResult> Chain([FromRoute] string ticker) =>
            this.OkOrError(
                handler.Handle(
                    new ChainQuery(
                        ticker: new Ticker(ticker), userId: User.Identifier()
                    )
                )
            );
        
        [HttpGet("export")]
        public async Task<ActionResult> Export()
        {
            return this.GenerateExport(
                await handler.Handle(
                    new ExportQuery(
                        User.Identifier()
                    )
                )
            );
        }
    }
}

File: src/web.fs/Controllers/OptionsController.fs
namespace web.Controllers

open System.Threading.Tasks
open Microsoft.AspNetCore.Mvc
open Microsoft.AspNetCore.Authorization
open core.fs.Options
open core.Shared
open web.Utils

[<ApiController>]
[<Authorize>]
[<Route("api/[controller]")>]
type OptionsController(handler: OptionsHandler) =
    inherit ControllerBase()

    [<HttpGet("pricing")>]
    member this.GetOptionPricing([<FromQuery>] symbol: string) =
        this.OkOrError(
            handler.Handle
                {
                    OptionPricingQuery.UserId = this.User.Identifier()
                    Symbol = OptionTicker.create(symbol)
                }
        )

    [<HttpGet("chain/{ticker}")>]
    member this.Chain([<FromRoute>] ticker: string) =
        this.OkOrError(
            handler.Handle
                {ChainQuery.Ticker = Ticker(ticker); UserId = this.User.Identifier()}
        )

    [<HttpGet("export")>]
    member this.Export() = task {
        let! result = handler.Handle {ExportQuery.UserId = this.User.Identifier()}
        return this.GenerateExport result
    }
